Jetpack Licensing: Streamlined logic for license issuing and payment method requirements

heydemoura opened the pull request:
#### Proposed Changes

* Context: 1201801459945917-as-1202552164778579/f
* Implemented a long term solution for the license issuing process with the issue license step coming before adding a payment method;
* The license issuing process is now more streamlined and does not rely only on a single component.

**Regarding `useLicenseIssuing(selectedSite?, product?)` Hook**
* Extracted license issuing mutation logic from issue license form to a hook, which can be called when successfully adding a payment method;
* Renamed `hooks.ts` to `hooks.tsx`, it might look like a whole file of additions, but it is just a new hook;

**`IssueLicenseForm`**
* No longer issuing a license from the license issuing form if there's no valid payment method on file. Instead, the license is issued after successfully adding a payment method (It will issue the license and not move to the payment method addition form if there's already a credit card on file);
* Moved auxiliary functions `selectProductOptions` and `alphabeticallySortedProductOptions` to `partner-portal/utils` module;
* Cleanup of `IssueLicenseForm` component, so that license issuing logic is handled on `useLicenseIssuing` hook, leaving mostly UI rendering and events implementation.
* There is a new event being tracked, labeled as `calypso_partner_portal_issue_selection_submit` for tracking when a user selects a product to be licensed and moves on to issuing the license or being required to input a payment method.

**`PaymentMethodAdd`**
* When adding a payment method, the _Go Back_ button can go back to product selection screen if there's a product selection;
* When adding a payment method with a `product` argument in the URL, it will then issue a license for that product after successfully adding the payment method;


#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Expected flows described here: pdpAdu-v0-p2

#### Screenshots
**Success notice when issuing a license**
![image](https://user-images.githubusercontent.com/5550190/186732814-151740ca-487a-452c-8e2f-cbdc33a74639.png)


#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in [Simple](P9HQHe-k8-p2-p2), [Atomic](P9HQHe-jW-p2-p2), and [self-hosted Jetpack sites](PCYsg-g6b-p2-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Are we memoizing when appropriate (for expensive computations)? More info in [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md) and [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors)
- [ ] Have we sent any new strings [for translation](PCYsg-1vr-p2-p2) ASAP?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

heydemoura committed:
Move license issuing mutation definition to a hook

heydemoura committed:
Add callback to the hook to require payment method

heydemoura committed:
Add redirect for payment method requiriment

heydemoura committed:
Remove creating licenses from GET requests

heydemoura committed:
Fix typo

heydemoura committed:
Refactor to issue license when adding payment method

heydemoura committed:
Refactor to extract product state from hook to component

heydemoura committed:
Add conditional href to payment method addition back button

heydemoura committed:
Refactor to enable license assingment payment method requirement

heydemoura committed:
Move license issuing auxiliary funtions to `partner-portal/utils`;

heydemoura committed:
Update payment-method-add tests

heydemoura commented on client/jetpack-cloud/sections/partner-portal/hooks.tsx:
```diff
@@ -0,0 +1,281 @@
+		return page.redirect( '/dashboard' );
+	};
+
+	const requirePaymentMethod = useCallback( () => {
```
Exported callback to handle redirection to `/payment-methods/add` logic.

heydemoura commented on client/jetpack-cloud/sections/partner-portal/hooks.tsx:
```diff
@@ -0,0 +1,281 @@
+				);
+			}
+
+			page.redirect( nextStep );
```
Removed the case that the user is redirected to `/payment-methods/add` as the following step, since this is not a real case anymore. Using `requirePaymentMethod()` on the component level will lead the user to `/payment-methods/add` before issuing a license.

heydemoura commented on client/jetpack-cloud/sections/partner-portal/issue-license-form/index.tsx:
```diff
@@ -42,141 +29,17 @@ export default function IssueLicenseForm( {
-			return false;
-		},
-	} );
+	const [ issueLicense, isSubmitting, requirePaymentMethod ] = useLicenseIssuing(
```
The above was moved to the `useLicenseIssuing` hook

heydemoura commented on client/jetpack-cloud/sections/partner-portal/issue-license-form/index.tsx:
```diff
@@ -205,22 +68,15 @@ export default function IssueLicenseForm( {
-		issueLicense.mutate( { product } );
-	}, [ dispatch, product, issueLicense.mutate ] );
 
-	const selectedSiteDomian = selectedSite?.domain;
```
Fixed a typo

heydemoura commented on client/jetpack-cloud/sections/partner-portal/issue-license-form/index.tsx:
```diff
@@ -205,22 +68,15 @@ export default function IssueLicenseForm( {
+		if ( hasPaymentMethod ) {
+			issueLicense.mutate( { product } );
+		} else {
+			requirePaymentMethod();
```
If there's a payment method, it will issue the license already.

heydemoura commented on client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx:
```diff
@@ -59,11 +64,18 @@ function PaymentMethodAdd(): ReactElement {
-	useReturnUrl( hasPaymentMethod );
+	const [ issueLicense, isSubmitting ] = useLicenseIssuing( null, product );
+
+	useReturnUrl( !! returnQueryArg && hasPaymentMethod );
```
Only redirects in case there's a `return` query argument. So it will not redirect on the regular case, that there's only a `product` query argument or no argument at all. The `return` query argument will only be present when the user tries to assign a license without a payment method on file.

atanas-dev replied:
useReturnUrl() already checks if the return query arg is empty so `!! returnQueryArg` is unnecessary I think.

heydemoura commented on client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx:
```diff
@@ -96,7 +108,7 @@ function PaymentMethodAdd(): ReactElement {
 
 	const successCallback = useCallback( () => {
-		if ( returnQueryArg ) {
+		if ( returnQueryArg || product ) {
```
Refetch the cards to update the `has_valid_payment_method` flag when a product is selected.

atanas-dev replied:
If I'm understanding this correctly, if there's only `product` but no `returnQueryArg` then we have nowhere to return to and refetching the cards to update `has_valid_payment_method` seems unnecessary since we won't redirect the user even if they have a new payment method i.e.  we should be fine with removing `|| product`.

heydemoura commented on client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx:
```diff
@@ -106,7 +118,13 @@ function PaymentMethodAdd(): ReactElement {
+		if ( product && hasPaymentMethod ) {
+			issueLicense.mutate( { product } );
+		}
+	}, [ hasPaymentMethod, product ] );
```
Issue the license once there's a product and a payment method on file.

heydemoura commented on client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx:
```diff
@@ -186,8 +206,12 @@ function PaymentMethodAdd(): ReactElement {
+											? addQueryArgs( { product }, partnerPortalBasePath( '/issue-license' ) )
+											: '/partner-portal/payment-methods/'
+									}
+									disabled={ isStripeLoading || issueLicense.isLoading || isSubmitting }
```
"Go Back" button will lead to the products list with the selected product highlighted, if there's a product selected, keeping the user on the license issuing flow.

heydemoura commented on client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/test/payment-method-add.jsx:
```diff
@@ -3,24 +3,30 @@
+				<Provider store={ store }>
+					<PaymentMethodAdd />
+				</Provider>
+			</QueryClientProvider>
```
Now that licenses can be issued from the this component, these mocks were needed for testing.

heydemoura commented on client/jetpack-cloud/sections/partner-portal/utils.ts:
```diff
@@ -176,3 +182,13 @@ export function formatApiPartner( partner: APIPartner ): Partner {
+	families: APIProductFamily[]
+): APIProductFamilyProduct[] {
+	return sortBy( selectProductOptions( families ), ( product ) => product.name );
+}
```
Auxiliary functions for license issuing have been moved from the component level to this utils module.

heydemoura commented on client/state/partner-portal/licenses/hooks/use-issue-license-mutation.ts:
```diff
@@ -2,7 +2,7 @@ import { useMutation, UseMutationOptions, UseMutationResult } from 'react-query'
-interface MutationIssueLicenseVariables {
+export interface MutationIssueLicenseVariables {
 	product: string;
 }
```
Exporting interface for proper typing on the hook level

atanas-dev commented on client/jetpack-cloud/sections/partner-portal/hooks.tsx:
```diff
@@ -0,0 +1,281 @@
+	const translate = useTranslate();
+	const dispatch = useDispatch();
+	const sites = useSelector( getSites ).length;
+	const [ isSubmitting, setIsSubmitting ] = useState( false );
```
Since we're modifying this, let's replace isSubmitting with a constant rather than state as its duplicating state that react-query already provides us with:
```
const isLoading = issueLicense.isLoading || assignLicense.isLoading.
```
using isLoading for consistency with react-query's naming.

atanas-dev commented on client/jetpack-cloud/sections/partner-portal/hooks.tsx:
```diff
@@ -0,0 +1,281 @@
+	} );
+
+	const paymentMethodRequired = useSelector( doesPartnerRequireAPaymentMethod );
+	const hasPaymentMethod = useSelector( hasValidPaymentMethod );
```
`doesPartnerRequireAPaymentMethod` already covers `hasPaymentMethod` so we can skip it.

atanas-dev commented on client/jetpack-cloud/sections/partner-portal/hooks.tsx:
```diff
@@ -0,0 +1,281 @@
+							)
+						);
+						return;
+					}
```
If we no longer have the case where we try to issue a license without being sure if a payment method is available, we should probably remove this logical branch.

atanas-dev commented on client/jetpack-cloud/sections/partner-portal/issue-license-form/index.tsx:
```diff
@@ -205,22 +68,15 @@ export default function IssueLicenseForm( {
-				removeQueryArgs( window.location.pathname + window.location.search, 'product' )
-			);
-			issueLicense.mutate( { product: defaultProduct } );
+		if ( hasPaymentMethod ) {
```
We should use `! paymentMethodRequired` here instead otherwise this fails for non-agencies that don't have a payment method as they go the requirePaymentMethod() route which does nothing since a payment method is not required for non-agencies (i.e. non-agencies can't issue licenses right now :D).

atanas-dev commented on client/jetpack-cloud/sections/partner-portal/issue-license-form/index.tsx:
```diff
@@ -205,22 +68,15 @@ export default function IssueLicenseForm( {
 
 	const onIssueLicense = useCallback( () => {
 		dispatch( recordTracksEvent( 'calypso_partner_portal_issue_license_submit', { product } ) );
```
We may want to move this even to the hook that handles issuing since we'd lose the event when the issuing happens when adding a payment method and issuing a license in one step.

heydemoura replied:
Smart! I've moved it to the hook level and replaced this event with `calypso_partner_portal_issue_selection_submit` for good measure, then we know if the user has moved on a step further.

atanas-dev commented on client/jetpack-cloud/sections/partner-portal/hooks.tsx:
```diff
@@ -0,0 +1,281 @@
+		},
+	} );
+
+	return [ issueLicense, isSubmitting, requirePaymentMethod ];
```
I think we should restructure this return value a bit like so:
```
const issue = useCallback( () => {
	if ( isLoading ) {
		return;
	}

	if ( paymentMethodRequired ) {
		const nextStep = addQueryArgs(
			{
				product,
			},
			partnerPortalBasePath( '/payment-methods/add' )
		);
		page( nextStep );
		return;
	}

	issueLicense.mutate( { product } );
}, [ product, paymentMethodRequired, issueLicense.mutate ] );

return [ issue, isLoading ];
```

This way callers of this hook don't have to worry about the payment method status and execute extra redirection functions - they only need to care about `isLoading` and to call `issue()`.
(isLoading comes from my earlier comment about replacing isSubmitting).

heydemoura replied:
Yes, this is a far better approach. It reduces verbosity on every place we would need to issue a license.

atanas-dev commented on client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx:
```diff
@@ -186,8 +206,12 @@ function PaymentMethodAdd(): ReactElement {
+									href={
+										product
+											? addQueryArgs( { product }, partnerPortalBasePath( '/issue-license' ) )
+											: '/partner-portal/payment-methods/'
```
Let's use `partnerPortalBasePath()` for new references to partner portal URLs i.e. `partnerPortalBasePath( '/payment-methods' )`.

Let's also move this conditional before rendering to make rendering easier to read.

atanas-dev commented on client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx:
```diff
@@ -186,8 +206,12 @@ function PaymentMethodAdd(): ReactElement {
-									disabled={ isStripeLoading }
+									href={
+										product
+											? addQueryArgs( { product }, partnerPortalBasePath( '/issue-license' ) )
```
Props for retaining some benefits from the last-minute implementation we had before so we could pre-select a product for the customer when they go back!

atanas-dev commented on client/jetpack-cloud/sections/partner-portal/utils.ts:
```diff
@@ -176,3 +182,13 @@ export function formatApiPartner( partner: APIPartner ): Partner {
+	return families.flatMap( ( family ) => family.products );
+}
+
+export function alphabeticallySortedProductOptions(
```
Let's rename this to make it obvious it's a query selector function like the one above:
`selectAlphabeticallySortedProductOptions`

atanas-dev reviewed (changes requested):
Thank you for taking the lead on fixing our short-term cheating! 😆
Changes mostly look good - I've sprinkled comments around places where we can improve or simplify things.

Note that some of my comments may be on things that were just moved but since we're touching them now we might as well add some minor improvements to them.

heydemoura committed:
Refactor license issuing, cleanup of old logic

heydemoura committed:
Refactor to improve component logic and tracking

heydemoura committed:
Fix payment method check middleware wrong redirect

Valid payment method check middleware was redirecting not to the page
that the user was navigating to, but to the one they were at.

heydemoura committed:
Refactor license issue hook for simpler logic

heydemoura committed:
Refactor license issuing form to work w/ new hook

heydemoura committed:
Rename utilitty function for clearer understanding

heydemoura committed:
Refactor payment method add form removing unecessary logic

atanas-dev commented on client/jetpack-cloud/sections/partner-portal/controller.tsx:
```diff
@@ -242,10 +242,10 @@ export function requireSelectedPartnerKeyContext(
 
 	if ( paymentMethodRequired ) {
-		const returnUrl = ensurePartnerPortalReturnUrl( pathname + search );
+		const returnUrl = ensurePartnerPortalReturnUrl( path );
```
Is this change necessary? Ideally, this would redirect back to whatever the current URL was regardless of its query params.

heydemoura replied:
There was a bug that was fixed with other changes. So I'm reverting that change.

atanas-dev commented on client/jetpack-cloud/sections/partner-portal/hooks.tsx:
```diff
@@ -0,0 +1,269 @@
+	const isLoading = useMemo(
+		() => assignLicense.isLoading || issueLicense.isLoading,
+		[ issueLicense, assignLicense ]
+	);
```
This is overkill - `const isLoading = assignLicense.isLoading || issueLicense.isLoading;` is sufficient and faster since we're testing two booleans instead of involving function calls, memoization etc.

atanas-dev commented on client/jetpack-cloud/sections/partner-portal/hooks.tsx:
```diff
@@ -0,0 +1,269 @@
+		}
+
+		issueLicense.mutate( { product } );
+	}, [ product, paymentMethodRequired, issueLicense.mutate ] );
```
```suggestion
	}, [ isLoading, product, paymentMethodRequired, issueLicense.mutate ] );
```

atanas-dev commented on client/jetpack-cloud/sections/partner-portal/hooks.tsx:
```diff
@@ -0,0 +1,269 @@
+
+	const issueLicense = useIssueLicenseMutation( {
+		onSuccess: ( license ) => {
+			dispatch( recordTracksEvent( 'calypso_partner_portal_issue_license_submit', { product } ) );
```
Let's rename this to something like `calypso_partner_portal_license_issued ` since this is triggered after a license is successfully issued.

atanas-dev commented on client/jetpack-cloud/sections/partner-portal/issue-license-form/index.tsx:
```diff
@@ -204,23 +62,11 @@ export default function IssueLicenseForm( {
-		dispatch( recordTracksEvent( 'calypso_partner_portal_issue_license_submit', { product } ) );
-		issueLicense.mutate( { product } );
-	}, [ dispatch, product, issueLicense.mutate ] );
+		dispatch( recordTracksEvent( 'calypso_partner_portal_issue_selection_submit', { product } ) );
```
How about we move this to the hook that returns the issue function e.g.:
```
const issue = useCallback( () => {
	if ( isLoading ) {
		return;
	}
	dispatch( recordTracksEvent( 'calypso_partner_portal_issue_license_submit', { product } ) );
	// ...
```
This way the hook will handle this and not whatever is calling it every time.

Bonus: After this change we can remove the entire `onIssueLicense` handler and pass issueLicense to onClick directly.

atanas-dev commented on client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx:
```diff
@@ -127,7 +145,9 @@ function PaymentMethodAdd(): ReactElement {
 			<SidebarNavigation />
 
-			{ returnQueryArg && <AssignLicenseStepProgress currentStep="addPaymentMethod" /> }
+			{ ( !! returnQueryArg || product ) && (
```
```suggestion
			{ ( !! returnQueryArg && !! product ) && (
```
I think we need both in order to consider this a part of the steps since no return = no flow / no next step and no product = no license issued.

heydemoura replied:
Not exactly, they happen in different situations, not always TRUE together, and for those cases we would want to have the progress bar being rendered.

1. User has previously issued a license, deleted the payment methods, went to assign that license and was sent to payment method input. In this case there is no `product` parameter, just `return`, it would need to show the progress bar, as they are seemingly resuming the process.
2. User is issuing a license. There is no `return` parameter as they will be sent to a next step as a regular issuing process, but there's the `product` parameter.

atanas-dev replied:
Makes sense 👍

atanas-dev commented on client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx:
```diff
@@ -104,9 +116,15 @@ function PaymentMethodAdd(): ReactElement {
+	}, [ page, product, issueLicense ] );
+
+	useEffect( () => {
+		if ( product && paymentMethodRequired ) {
```
This doesn't work as expected - adding a payment method while issuing a license redirects to the payment methods listing page instead of to the assign license step.

After some debugging it seems this works:
```
const successCallback = useCallback( () => {
	if ( product ) {
		// Refresh stored cards to establish whether there is a valid payment method now.
		reduxDispatch(
			fetchStoredCards( {
				startingAfter: '',
				endingBefore: '',
			} )
		);
	} else {
		page( partnerPortalBasePath( '/payment-methods' ) );
	}
}, [ page, product, reduxDispatch ] );

useEffect( () => {
	if ( product && ! paymentMethodRequired ) {
		issueLicense();
	}
}, [ product, paymentMethodRequired ] );
```

The key changes here are:
- `if ( returnQueryArg ) {` -> `if ( product ) {`
- `if ( product && paymentMethodRequired ) {` -> `if ( product && ! paymentMethodRequired ) {`

Note that even with this change the process is a bit janky:
- The Go Back button is enabled during the process.
- We don't show a success notice for the license being issued - IMO we should make it clear that a license was issued since they are going to be charged for it even if they bail from the process before assigning the license.

We may want to address this in a separate task to make the experience smoother.

heydemoura replied:
Fixed the conditionals with the suggested changes, the redirect now works as expected.

heydemoura replied:
The process is a bit janky, we should make a separate task to make that smoother, which will involve less code and objective PRs for specific improvements.

heydemoura replied:
Added a success notice for when the user successfully issues a license.

heydemoura replied:
I realized that we would need to test both `returnQueryArg` or `product`, as there's the case when a user has issued a license previously, had removed the payment method or it got expired, and will resume the process clicking on `Assign License` on the licenses list. This would send the user to the `payment-methods/add` page with a `return` query parameter, on that case we should not send the user to `/partner-portal/payment-methods` but send him to the license assignment (which is on the `return` query argument). Similar to what we have for rendering the progress bar.

So the approach would be:

```javascript
	const successCallback = useCallback( () => {
		if ( returnQueryArg || product ) {
			reduxDispatch(
				fetchStoredCards( {
					startingAfter: '',
					endingBefore: '',
				} )
			);
		} else {
			page( partnerPortalBasePath( '/payment-methods' ) );
		}
	}, [ page, product, reduxDispatch, partnerPortalBasePath, returnQueryArg ] );
```

atanas-dev replied:
Nice catch 👍

atanas-dev reviewed (changes requested):
Things look a lot neater right now - props!

I've noted a few improvements but the major blocking piece is that the flow is broken right now as the user is shown the payment methods list instead of the assign license step after adding a payment method.

---

FYI I believe a12s will still be required to have a payment method because of a change introduced by Genesis where if the WPCOM user has an agency partner associated with it, even if that partner is not the current partner, the payment method requirement will apply because of how isAgencyUser() works.
If we decide to address this we can do so in a separate patch as it's not necessarily related to this one.

heydemoura committed:
Remove overkill memoization

heydemoura committed:
Update license issuing tracked event name

heydemoura committed:
Update moving tracking from component to hook level

heydemoura committed:
Fix conditionals on successful payment method addition

heydemoura committed:
Add notice informing which license was issued

heydemoura committed:
Refactor success of payment method addition to resunme the flow

heydemoura committed:
Revert changes due no longer reproducible bug

There was a bug that was redirecting the user to payment methods
but that seemed to be fixed with other changes

atanas-dev reviewed (approved):
Patch tests good for me - great work! 🚢

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7495876

Hi @heydemoura, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include this string: `Your %(product)s license was successfuly issued `

Thank you in advance!

emilyaudela commented:
Hi there! It looks like there is a typo is this [string](https://github.com/Automattic/wp-calypso/blob/trunk/client/jetpack-cloud/sections/partner-portal/hooks.tsx#L201)

`Your %(product)s license was successfuly issued`

"sucessfuly" should be "successfully"

heydemoura commented:
> Hi there! It looks like there is a typo is this [string](https://github.com/Automattic/wp-calypso/blob/trunk/client/jetpack-cloud/sections/partner-portal/hooks.tsx#L201):
> 
> `Your %(product)s license was successfuly issued`
> 
> "sucessfuly" should be "successfully"

Thanks for spotting this,
This typo is being handled on a different PR.
https://github.com/Automattic/wp-calypso/pull/67044

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7495876).
